### PR TITLE
Auth config and USER CRUD

### PR DESCRIPTION
**Pull request by Georgio-Khoury**

### CODE REVIEW

File: packages/backend/drizzle.config.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ export default defineConfig({
   out: './src/db/migrations',
   dialect: 'postgresql',
   dbCredentials: {
-    url: 'postgresql://postgres:letmein@localhost:5432/Develeb-platform',
+    url: process.env.DATABASE_URL!,
```
Comment (Hkataya) [inline packages/backend/drizzle.config.ts:10]: why the !

File: packages/backend/src/api/user/userRepository.ts [hunk 2]
```diff
@@ -1,16 +1,87 @@
...
+import { eq } from 'drizzle-orm';
+import { db } from 'src/db';
+import { user } from 'src/db/schema';
 
-export const users: User[] = [
-  { id: 1, name: 'Alice', email: 'alice@example.com', age: 42, createdAt: new Date(), updatedAt: new Date() },
-  { id: 2, name: 'Bob', email: 'bob@example.com', age: 21, createdAt: new Date(), updatedAt: new Date() },
-];
+import { User } from '../user/userModel';
+// export const users: User[] = [
```
Comment (Hkataya) [inline packages/backend/src/api/user/userRepository.ts:8]: is there a reason for keeping the comments?

File: packages/backend/src/api/user/userRepository.ts [hunk 3]
```diff
@@ -1,16 +1,87 @@
...
+//     level_id: 1,
+//     phone_number: '+961 54738292',
+//     tags: 'developer',
+//   },
+// ];
 
 export const userRepository = {
-  findAllAsync: async (): Promise<User[]> => {
-    return users;
+  findAllAsync: async () => {
```
Comment (Hkataya) [inline packages/backend/src/api/user/userRepository.ts:46]: Where's the type?

File: packages/backend/src/api/user/userRouter.ts [hunk 4]
```diff
@@ -13,15 +20,15 @@ userRegistry.register('User', UserSchema);
...
+  router.use(bodyParser.json());
   userRegistry.registerPath({
     method: 'get',
     path: '/users',
     tags: ['User'],
     responses: createApiResponse(z.array(UserSchema), 'Success'),
   });
 
-  router.get('/', async (_req: Request, res: Response) => {
+  router.get('/', authenticate, authorizeRole('admin'), async (_req: Request, res: Response) => {
```
Comment (Hkataya) [inline packages/backend/src/api/user/userRouter.ts:31]: The role should not be hard coded

File: packages/backend/src/api/user/userRouter.ts [hunk 5]
```diff
@@ -34,11 +41,93 @@ export const userRouter: Router = (() => {
...
+  });
+
+  router.delete('/:id', authenticate, async (req: Request, res: Response) => {
+    const id = req.params.id;
+    if (req.user && req.user.id === id) {
+      //making sure that the authenticated user is deleting his own account and not someone else's account
+      const serviceResponse = await userService.deleteUser(id);
+      handleServiceResponse(serviceResponse, res);
+    } else {
+      res.status(401).json({ message: 'Unauthorized' }); //do like this or create a serviceResponse object, assign it the error code and message and send it to handleserviceresponse?
```
Comment (Hkataya) [inline packages/backend/src/api/user/userRouter.ts:72]: Better to have a serviceResponse object

File: packages/backend/src/api/user/userRouter.ts [hunk 6]
```diff
@@ -34,11 +41,93 @@ export const userRouter: Router = (() => {
...
+    if (req.user && req.user.id === id) {
+      const serviceResponse = await userService.updateUser(id, full_name, level_id, category_id, tags);
+      handleServiceResponse(serviceResponse, res);
+    } else {
+      res.status(401).json({ message: 'Unauthorized' }); //do like this or create a serviceResponse object, assign it the error code and message and send it to handleserviceresponse?
+    }
+  });
+
+  router.post('/login', async (req, res, next) => {
+    passport.authenticate('local', { session: false }, (err, user, info) => {
```
Comment (Hkataya) [inline packages/backend/src/api/user/userRouter.ts:87]: We need to move the authenticate logic to a different service. The routes/controllers should not be responsible for auth logic.

File: packages/backend/src/api/user/userRouter.ts [hunk 7]
```diff
@@ -34,11 +41,93 @@ export const userRouter: Router = (() => {
...
+        const token = jwt.sign({ id: user[0].id, role: user[0].role }, env.JWT_SECRET, { expiresIn: '1h' });
+
+        const refreshToken = jwt.sign({ id: user[0].id, role: user[0].role }, env.JWT_REFRESH_SECRET, { expiresIn: '7d' });
+        return res.json({ message: 'Login successful', token, refreshToken });
+      });
+    })(req, res, next);
+  });
+
+  router.post('/refresh-token', async (req, res) => {
+    const { refreshToken } = req.body;
```
Comment (Hkataya) [inline packages/backend/src/api/user/userRouter.ts:108]: Same here, move logic to an auth service

File: packages/backend/src/api/user/userService.ts [hunk 8]
```diff
@@ -35,4 +35,67 @@ export const userService = {
...
+    } catch (ex) {
+      const errorMessage = `Error creating user: ${(ex as Error).message}`;
+      logger.error(errorMessage);
+      return new ServiceResponse(ResponseStatus.Failed, errorMessage, null, StatusCodes.INTERNAL_SERVER_ERROR);
+    }
+  },
+
+  deleteUser: async (id: string) => {
+    try {
+      console.log(id);
```
Comment (Hkataya) [inline packages/backend/src/api/user/userService.ts:73]: We should not keep console.logs in the code.

File: packages/backend/src/api/user/userService.ts [hunk 9]
```diff
@@ -35,4 +35,67 @@ export const userService = {
...
+    phone_number: string,
+    level_id: number,
+    category_id: number
+  ) => {
+    try {
+      const user = await userRepository.findByEmailAsync(email);
+      if (user.length !== 0) {
+        return new ServiceResponse(ResponseStatus.Failed, 'User already exists', null, StatusCodes.CONFLICT);
+      }
+      const newUser = await userRepository.createUserAsync(
```
Comment (Hkataya) [inline packages/backend/src/api/user/userService.ts:54]: We need to use the logger object for logging even successful operations

**Review by Hkataya — CHANGES_REQUESTED** [review]
Can you write tests for this?

File: packages/backend/drizzle.config.ts (hunk 1 above)
Comment (Georgio-Khoury) [reply packages/backend/drizzle.config.ts:10]: The type of the process.env.DATABASE_URL is string|undefined which is generating the error saying url can only be a string and not string|undefined. The ! serves to tell the compiler that this value cannot be undefined.

File: packages/backend/src/api/user/userRepository.ts (hunk 3 above)
Comment (Hkataya) [reply packages/backend/src/api/user/userRepository.ts:46]: Pagination

**Review by Hkataya — CHANGES_REQUESTED** [review]

File: packages/backend/src/api/user/userRouter.ts [hunk 10]
```diff
@@ -1,44 +1,113 @@
...
+} from './userRequest';
 
 export const userRegistry = new OpenAPIRegistry();
 
 userRegistry.register('User', UserSchema);
 
 export const userRouter: Router = (() => {
   const router = express.Router();
-
+  router.use(bodyParser.json());
```
Comment (Hkataya) [inline packages/backend/src/api/user/userRouter.ts:37]: why do we need this? we already have app.use(express.json()); in the server.ts file

Comment (Hkataya) [reply packages/backend/src/api/user/userRouter.ts:37]: https://stackoverflow.com/questions/47232187/express-json-vs-bodyparser-json

File: packages/backend/tsconfig.json [hunk 11]
```diff
@@ -16,5 +16,4 @@
   },
   "extends": [],
   "exclude": ["node_modules"],
-  "include": ["src/**/*.ts"]
```
Comment (Hkataya) [inline packages/backend/tsconfig.json:19]: why remove this

File: packages/backend/src/common/utils/paginationHandler.ts [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import { like } from 'drizzle-orm';
+import { PgTableWithColumns } from 'drizzle-orm/pg-core';
+
+export const queryFilters = (entity: PgTableWithColumns<any>, filters: { [key: string]: string | undefined }) => {
```
Comment (Hkataya) [inline packages/backend/src/common/utils/paginationHandler.ts:4]: let's just keep the queryfilters in the repository. utils should be generic and not depend on a specific database connector

**Review by Hkataya — CHANGES_REQUESTED** [review]

**Review by Hkataya — APPROVED** [review]

**Pull request merged**